Engagement Workflow v2

atGit2021 opened the pull request:
[Monday](https://seed-company-squad.monday.com/boards/3451697530/views/78801638/pulses/6767409800)

Commit:
Engagement Workflow v2

Commit:
Engagement Workflow v2

atGit2021 commented:
Todo: commented out edgedb related code for now; needs engagement workflow schema

atGit2021 commented:
@CarsonF I thought I would have you review before starting to dig into the migration file.

CarsonF commented on src/components/engagement/engagement.service.ts:
```diff
@@ -288,9 +279,17 @@ export class EngagementService {
       session,
     );
 
+    let updated = previous;
```
Doesn't look like this is needed
```suggestion
```

CarsonF commented on src/components/engagement/workflow/engagement-workflow.repository.ts:
```diff
@@ -0,0 +1,69 @@
+    engagementId: ID<'Engagement'>,
+    steps: readonly EngagementStatus[],
+  ) {
+    const query = edgeql(`
+        with
+         engagement := <Engagement><uuid>$engagementId,
+         steps := array_unpack(<array<Engagement::Status>>$steps),
+         mostRecentEvent := (
+          select engagement.workflowEvents
+          filter .to in steps if exists steps else true
+          order by .at desc
+          limit 1
+        )
+        select mostRecentEvent.to
+      `);
```
```suggestion
    const query = edgeql(`
      with
       engagement := <Engagement><uuid>$engagementId,
       steps := array_unpack(<array<Engagement::Status>>$steps),
       mostRecentEvent := (
        select engagement.workflowEvents
        filter .to in steps if exists steps else true
        order by .at desc
        limit 1
      )
      select mostRecentEvent.to
    `);
```

atGit2021 replied:
@CarsonF  Sorry, just not seeing the difference in your suggested comment. What's the suggested change(s) on this specific query?

CarsonF replied:
Reduce the indentation, so the ending backtick is on the same level and then inner body is indented once.

CarsonF commented on src/components/engagement/workflow/engagement-workflow.service.ts:
```diff
@@ -0,0 +1,127 @@
+  ) {
+    const { engagement: engagementId, notes } = input;
+
+    const { EngagementLoader } = await import('../engagement.loader');
+    const engagements = await this.resources.getLoader(EngagementLoader);
+    const loaderKey = {
+      id: engagementId,
+      view: { active: true },
+    } as const;
+    const previous = await engagements.load(loaderKey);
```
I did switch up this logic in project recently. Check out the new version.

atGit2021 replied:
@CarsonF, just out of curiosity, why did you decide to go the route of importing the service file to query the needed project (engagement in this case) instead of using the .Loader approach?  Is it just less lines of code or are we moving away from Loaders?

CarsonF commented on src/components/engagement/workflow/events/engagement-transitioned.event.ts:
Since we aren't publishing this event, we can remove this file

CarsonF reviewed (changes requested):
Looks really close other than a few minor things!

Drop 788f86354ce55b58cc449bfd7d6e09f9d51f63f7 - we aren't versioning these access policies since they are generated.

CarsonF commented on dbschema/engagement-workflow.esdl:
```diff
@@ -0,0 +1,66 @@
+    );
+  }
+
+  scalar type Status extending enum<
```
Don't forget to adjust order here to match app code.

CarsonF committed:
Rename step -> state

CarsonF committed:
Drop unused code

CarsonF committed:
Polish transition names

CarsonF committed:
Rename file

CarsonF committed:
Add project step condition

CarsonF committed:
Fix permissions

In general, just tried to follow suit of project workflow

CarsonF committed:
Remove end/reject proposal condition on project state

Seth agrees this doesn't make sense

CarsonF committed:
Work around needing engagement.project.step for migration

CarsonF committed:
Drop another superfluous override

Commit:
Revised engagement seeding for new engagement workflow events

CarsonF committed:
Remove most of the engagement workflow tests / Replace with a couple late start & early end tests

Mostly, we just want the engagement status to follow the project status.
This is already being tested in engagement.e2e-spec.ts.
So I removed all of these.

I did add a couple where we think users will be directly interacting with the engagement workflow

Commit:
publish event for engagements

atGit2021 commented on src/components/engagement/events/engagement-updated.event.ts:
```diff
@@ -11,8 +11,8 @@ export class EngagementUpdatedEvent {
     readonly previous: UnsecuredDto<Engagement>,
-    readonly input: UpdateLanguageEngagement | UpdateInternshipEngagement,
     readonly session: Session,
+    readonly input?: UpdateLanguageEngagement | UpdateInternshipEngagement,
```
@CarsonF I'm not sure if changing the input param to optional has further impacts but based on high level scans I think this will be ok, but let me know if there is a better way to design the event publishing for engagement workflow.
@bryanjnelson @willdch

Commit:
removed conflicting migration file

Commit:
add primary location to test project

Commit:
added funding account to project in order to resolve missing departmentId

Commit:
removed ProjectStep PendingTerminationApproval to Terminated from array

Commit:
fix permissions error when creating engagement

atGit2021 commented on test/engagement.e2e-spec.ts:
```diff
@@ -1071,11 +1071,11 @@ describe('Engagement e2e', () => {
-      ProjectStep.Terminated,
-      EngagementStatus.Terminated,
-    ],
+    // [
```
@CarsonF I commented out this element of the array which was causing the following error.  I'm not sure how to fix this, but putting my findings below:
```
 should update Engagement status to match Project step when it becomes PendingTerminationApproval
    [Unauthorized, Input, Client]: This transition is not available
```
at the following location: 
`at findTransition (src/components/workflow/workflow.service.ts:159:11)`

atGit2021 commented on test/engagement-workflow.e2e-spec.ts:
```diff
@@ -1,212 +1,85 @@
-      expect(lEngagementStatus.status.value).toBe(EngagementStatus.Active);
-    });
+    await controller.login();
+    await eng.executeByLabel('Approve');
```
@CarsonF , I executeByLabel('Approve') flows down to the method `findTransition` in the workflow.tester.ts file where it fails to find a transition and throws the exception, `This transition is not available`.  I tried to debug this and the rest of the CI errors, but I'm not understanding why the transitions are not being found.

atGit2021 commented on test/engagement-workflow.e2e-spec.ts:
```diff
@@ -1,212 +1,85 @@
-      );
-    });
+    await controller.login();
+    await eng.executeByState('Completed');
```
Similar exception described above is happening here.
